src/atomic.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2019 Alibaba Cloud Computing. All rights reserved.
+// Copyright (C) 2020 Red Hat, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A wrapper over an `ArcSwap<GuestMemory>` struct to support RCU-style mutability.
+//!
+//! With the `backend-atomic` feature enabled, simply replacing `GuestMemoryMmap`
+//! with `GuestMemoryAtomic<GuestMemoryMmap>` will enable support for mutable memory maps.
+//! To support mutable memory maps, devices will also need to use
+//! `GuestAddressSpace::memory()` to gain temporary access to guest memory.
+
+extern crate arc_swap;
+
+use arc_swap::{ArcSwap, Guard};
+use std::ops::Deref;
+use std::sync::{Arc, LockResult, Mutex, MutexGuard, PoisonError};
+
+use crate::{GuestAddressSpace, GuestMemory};
+
+/// A fast implementation of a mutable collection of memory regions.
+///
+/// This implementation uses ArcSwap to provide RCU-like snapshotting of the memory map:
+/// every update of the memory map creates a completely new GuestMemory object, and
+/// readers will not be blocked because the copies they retrieved will be collected once
+/// no one can access them anymore.  Under the assumption that updates to the memory map
+/// are rare, this allows a very efficient implementation of the `memory()` method.
+#[derive(Clone, Debug)]
+pub struct GuestMemoryAtomic<M: GuestMemory> {
+    // GuestAddressSpace<M>, which we want to implement, is basically a drop-in
+    // replacement for &M.  Therefore, we need to pass to devices the GuestMemoryAtomic
+    // rather than a reference to it.  To obtain this effect we wrap the actual fields
+    // of GuestMemoryAtomic with an Arc, and derive the Clone trait.  See the
+    // documentation for GuestAddressSpace for an example.
+    inner: Arc<(ArcSwap<M>, Mutex<()>)>,
+}
+
+impl<M: GuestMemory> From<Arc<M>> for GuestMemoryAtomic<M> {
+    /// create a new GuestMemoryAtomic object whose initial contents come from
+    /// the `map` reference counted GuestMemory.
+    fn from(map: Arc<M>) -> Self {
+        let inner = (ArcSwap::new(map), Mutex::new(()));
+        GuestMemoryAtomic {
+            inner: Arc::new(inner),
+        }
+    }
+}
+
+impl<M: GuestMemory> GuestMemoryAtomic<M> {
+    /// create a new GuestMemoryAtomic object whose initial contents come from
+    /// the `map` GuestMemory.
+    pub fn new(map: M) -> Self {
+        Arc::new(map).into()
+    }
+
+    fn load(&self) -> Guard<'static, Arc<M>> {
```

Comment:
Aren't we supposed to expose this one as public?

Reply:
Method load() will be hide behind GuestAddressSpace::memory(). External clients do not need direct access to the load method.

Reply:
Right, this method only hides the ugly `self.inner.0.`. `memory()` returns the correct wrapper for `arc_swap::Guard`.

Reply:
Yep I realized that after a few min :)